logbook/logbook.go

```diff
@@ -604,59 +604,115 @@ func (book *Book) WriteVersionDelete(ctx context.Context, initID string, revisio
 	return book.save(ctx)
 }
 
-// WritePublish adds an operation to a log marking the publication of a number
-// of versions to one or more destinations
-func (book *Book) WritePublish(ctx context.Context, initID string, revisions int, destinations ...string) error {
+// WriteRemotePush adds an operation to a log marking the publication of a
+// number of versions to a remote address. It returns a rollback function that
+// removes the operation when called
+func (book *Book) WriteRemotePush(ctx context.Context, initID string, revisions int, remoteAddr string) (l *oplog.Log, rollback func(ctx context.Context) error, err error) {
 	if book == nil {
-		return ErrNoLogbook
+		return nil, nil, ErrNoLogbook
 	}
-	log.Debugf("WritePublish: %s, revisions: %d, destinations: %v", initID, revisions, destinations)
+	log.Debugf("WriteRemotePush: %s, revisions: %d, remote: %q", initID, revisions, remoteAddr)
 
 	branchLog, err := book.branchLog(ctx, initID)
 	if err != nil {
-		return err
+		return nil, nil, err
 	}
 	if err := book.hasWriteAccess(branchLog.l); err != nil {
-		return err
+		return nil, nil, err
 	}
 
 	branchLog.Append(oplog.Op{
 		Type:      oplog.OpTypeInit,
-		Model:     PublicationModel,
+		Model:     PushModel,
 		Size:      int64(revisions),
-		Relations: destinations,
-		// TODO (b5) - finish
+		Relations: []string{remoteAddr},
 	})
 
-	return book.save(ctx)
+	if err = book.save(ctx); err != nil {
+		return nil, nil, err
+	}
+
+	// after successful save calling rollback drops the written push operation
+	// rollback can only be called once. it will overwrite itself with a noop func
+	// when called
+	rollback = func(ctx context.Context) error {
+		defer func() {
+			rollback = func(ctx context.Context) error { return nil }
```

Comment:
This is sweet!

Reply:
Damn. Super glad you caught this. I wrote a test for this exact purpose, and didn't compose the test properly.